README.md
<p align="center">
  <a href="https://github.com/node-minibase">
    <img height="250" width="250" src="https://avatars1.githubusercontent.com/u/23032863?v=3&s=250">
  </a>
</p>

# minibase-tests [![NPM version](https://img.shields.io/npm/v/minibase-tests.svg?style=flat)](https://www.npmjs.com/package/minibase-tests) [![NPM downloads](https://img.shields.io/npm/dm/minibase-tests.svg?style=flat)](https://npmjs.org/package/minibase-tests) [![npm total downloads][downloads-img]][downloads-url]

> Tests for applications built on [minibase][] or [base][]. All Base apps passes these tests.

[![code climate][codeclimate-img]][codeclimate-url] 
[![standard code style][standard-img]][standard-url] 
[![linux build status][travis-img]][travis-url] 
[![windows build status][appveyor-img]][appveyor-url] 
[![coverage status][coveralls-img]][coveralls-url] 
[![dependency status][david-img]][david-url]

## Table of Contents
- [Install](#install)
- [Usage](#usage)
- [API](#api)
  * [suite](#suite)
- [Related](#related)
- [Contributing](#contributing)
- [Building docs](#building-docs)
- [Running tests](#running-tests)
- [Author](#author)
- [License](#license)

_(TOC generated by [verb](https://github.com/verbose/verb) using [markdown-toc](https://github.com/jonschlinkert/markdown-toc))_

## Install
Install with [npm](https://www.npmjs.com/)

```
$ npm install minibase-tests --save
```

or install using [yarn](https://yarnpkg.com)

```
$ yarn add minibase-tests
```

## Usage
> For more use-cases see the [tests](test.js)

```js
const minibaseTests = require('minibase-tests')
```

## API

### [suite](index.js#L78)
> Test any app based on [minibase][] and [base][], just pass constructor as `App` argument. If it is `base` based pass `opts.isBase: true` option. When run `.runTests` it returns resolved Promise with array with length of 0 if all tests are passed. If any of the tests fails that `result` array will contain these tests - their title, index and the error. Resolved array also has `.tests` property which is the count of all tests, so easily you can do `res.tests - res.length` to find how many tests are failed, and get them by outputing `res`.

**Params**

* `App` **{Function}**: app constructor, if not a function returns rejected promise    
* `opts` **{Object}**: optional object, pass `isBase: true` for [base][] apps    
* `returns` **{Promise}**: promise if `App` not a function or instance of [Runner](./runner.js), so call `.runTests()`  

**Example**

```js
var suite = require('minibase-tests')
var Base = require('base')
var Assemble = require('assemble-core')
var Templates = require('templates')
var MiniBase = require('minibase').MiniBase

suite(Base, { isBase: true })
.runTests().then(function (res) {
  // if `res` has bigger length
  // it will contain failed tests
  console.log(res.length) // => 0
  console.log(res.tests) // => 17
})
suite(Assemble, { isBase: true })
.runTests().then(function (res) {
  console.log(res.length) // => 0
  console.log(res.tests) // => 17
})
suite(Templates, { isBase: true })
.runTests().then(function (res) {
  console.log(res.length) // => 0
  console.log(res.tests) // => 17
})

// MiniBase itself passes these tests too
suite(MiniBase).runTests().then(function (res) {
  console.log(res.length) // => 0
  console.log(res.tests) // => 17
})

function MyApp () {
  MiniBase.call(this)
}
MiniBase.extend(MyApp)

suite(MyApp).runTests().then(function (res) {
  console.log(res.length) // => 0
  console.log(res.tests) // => 17
})
```

## Related
- [assemble-core](https://www.npmjs.com/package/assemble-core): The core assemble application with no presets or defaults. All configuration is left to the implementor. | [homepage](https://github.com/assemble/assemble-core "The core assemble application with no presets or defaults. All configuration is left to the implementor.")
- [assemble](https://www.npmjs.com/package/assemble): Get the rocks out of your socks! Assemble makes you fast at creating web projects. Assemble… [more](https://github.com/assemble/assemble) | [homepage](https://github.com/assemble/assemble "Get the rocks out of your socks! Assemble makes you fast at creating web projects. Assemble is used by thousands of projects for rapid prototyping, creating themes, scaffolds, boilerplates, e-books, UI components, API documentation, blogs, building websit")
- [generate](https://www.npmjs.com/package/generate): Command line tool and developer framework for scaffolding out new GitHub projects. Generate offers the robustness… [more](https://github.com/generate/generate) | [homepage](https://github.com/generate/generate "Command line tool and developer framework for scaffolding out new GitHub projects. Generate offers the robustness and configurability of Yeoman, the expressiveness and simplicity of Slush, and more powerful flow control and composability than either.")
- [minibase-create-plugin](https://www.npmjs.com/package/minibase-create-plugin): Utility for [minibase][] and [base][] that helps you create plugins | [homepage](https://github.com/node-minibase/minibase-create-plugin#readme "Utility for [minibase][] and [base][] that helps you create plugins")
- [minibase-is-registered](https://www.npmjs.com/package/minibase-is-registered): Plugin for [minibase][] and [base][], that adds `isRegistered` method to your application to detect if plugin… [more](https://github.com/node-minibase/minibase-is-registered#readme) | [homepage](https://github.com/node-minibase/minibase-is-registered#readme "Plugin for [minibase][] and [base][], that adds `isRegistered` method to your application to detect if plugin is already registered and returns true or false if named plugin is already registered on the instance.")
- [minibase](https://www.npmjs.com/package/minibase): MiniBase is minimalist approach to Base - @node-base, the awesome framework. Foundation for building complex APIs… [more](https://github.com/node-minibase/minibase#readme) | [homepage](https://github.com/node-minibase/minibase#readme "MiniBase is minimalist approach to Base - @node-base, the awesome framework. Foundation for building complex APIs with small units called plugins. Works well with most of the already existing [base][] plugins.")
- [templates](https://www.npmjs.com/package/templates): System for creating and managing template collections, and rendering templates with any node.js template engine. Can… [more](https://github.com/jonschlinkert/templates) | [homepage](https://github.com/jonschlinkert/templates "System for creating and managing template collections, and rendering templates with any node.js template engine. Can be used as the basis for creating a static site generator or blog framework.")

## Contributing
Pull requests and stars are always welcome. For bugs and feature requests, [please create an issue](https://github.com/node-minibase/minibase-tests/issues/new).  
Please read the [contributing guidelines](CONTRIBUTING.md) for advice on opening issues, pull requests, and coding standards.

**In short:** If you want to contribute to that project, please follow these things

1. Please DO NOT edit [README.md](README.md), [CHANGELOG.md](CHANGELOG.md) and [.verb.md](.verb.md) files. See ["Building docs"](#building-docs) section.
2. Ensure anything is okey by installing the dependencies and run the tests. See ["Running tests"](#running-tests) section.
3. Always use `npm run commit` to commit changes instead of `git commit`, because it is interactive and user-friendly. It uses [commitizen][] behind the scenes, which follows Conventional Changelog idealogy.
4. Do NOT bump the version in package.json. For that we use `npm run release`, which is [standard-version][] and follows Conventional Changelog idealogy.

Thanks a lot! :)

## Building docs
Documentation and that readme is generated using [verb-generate-readme][], which is a [verb][] generator, so you need to install both of them and then run `verb` command like that

```
$ npm install verbose/verb#dev verb-generate-readme --global && verb
```

_Please don't edit the README directly. Any changes to the readme must be made in [.verb.md](.verb.md)._

## Running tests
Clone repository and run the following in that cloned directory

```
$ npm install && npm test
```

## Author
**Charlike Mike Reagent**

+ [github/tunnckoCore](https://github.com/tunnckoCore)
+ [twitter/tunnckoCore](http://twitter.com/tunnckoCore)

## License
Copyright © 2016, [Charlike Mike Reagent](http://www.tunnckocore.tk). Released under the [MIT license](LICENSE).

***

_This file was generated by [verb-generate-readme](https://github.com/verbose/verb-generate-readme), v0.2.0, on November 17, 2016._

[base]: https://github.com/node-base/base
[commitizen]: https://github.com/commitizen/cz-cli
[minibase]: https://github.com/node-minibase/minibase
[standard-version]: https://github.com/conventional-changelog/standard-version
[verb-generate-readme]: https://github.com/verbose/verb-generate-readme
[verb]: https://github.com/verbose/verb

[downloads-url]: https://www.npmjs.com/package/minibase-tests
[downloads-img]: https://img.shields.io/npm/dt/minibase-tests.svg

[codeclimate-url]: https://codeclimate.com/github/node-minibase/minibase-tests
[codeclimate-img]: https://img.shields.io/codeclimate/github/node-minibase/minibase-tests.svg

[travis-url]: https://travis-ci.org/node-minibase/minibase-tests
[travis-img]: https://img.shields.io/travis/node-minibase/minibase-tests/master.svg?label=linux

[appveyor-url]: https://ci.appveyor.com/project/tunnckoCore/minibase-tests
[appveyor-img]: https://img.shields.io/appveyor/ci/tunnckoCore/minibase-tests/master.svg?label=windows

[coveralls-url]: https://coveralls.io/r/node-minibase/minibase-tests
[coveralls-img]: https://img.shields.io/coveralls/node-minibase/minibase-tests.svg

[david-url]: https://david-dm.org/node-minibase/minibase-tests
[david-img]: https://img.shields.io/david/node-minibase/minibase-tests.svg

[standard-url]: https://github.com/feross/standard
[standard-img]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.1.2"></a>
## [1.1.2](https://github.com/node-minibase/minibase-tests/compare/v1.1.1...v1.1.2) (2016-11-17)


### Bug Fixes

* **package:** wildstar for minibase devDep ([e3b8eec](https://github.com/node-minibase/minibase-tests/commit/e3b8eec))



<a name="1.1.1"></a>
## [1.1.1](https://github.com/node-minibase/minibase-tests/compare/v1.1.0...v1.1.1) (2016-11-17)


### Bug Fixes

* **appveyor:** allow failures ([f2faf60](https://github.com/node-minibase/minibase-tests/commit/f2faf60))
* **tests:** fix and remove tests for "use" event ([6291e8f](https://github.com/node-minibase/minibase-tests/commit/6291e8f))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/node-minibase/minibase-tests/compare/v1.0.3...v1.1.0) (2016-11-14)


### Features

* **runTests:** better reporting, improve ".runTests" ([144fe78](https://github.com/node-minibase/minibase-tests/commit/144fe78))



<a name="1.0.3"></a>
## [1.0.3](https://github.com/node-minibase/minibase-tests/compare/v1.0.2...v1.0.3) (2016-11-14)


### Bug Fixes

* **test:** update one test ([ef4f201](https://github.com/node-minibase/minibase-tests/commit/ef4f201))



<a name="1.0.2"></a>
## [1.0.2](https://github.com/node-minibase/minibase-tests/compare/v1.0.1...v1.0.2) (2016-11-14)


### Bug Fixes

* **tests:** update failing tests ([c15efeb](https://github.com/node-minibase/minibase-tests/commit/c15efeb))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/node-minibase/minibase-tests/compare/v1.0.0...v1.0.1) (2016-11-14)


### Bug Fixes

* **package:** add runner.js to files ([7adf444](https://github.com/node-minibase/minibase-tests/commit/7adf444))





## 1.0.0 - 2016-11-14
- First release
- semantic versioning
- add docs
- update tests
- add tests
- implement

## 0.0.0 - 2016-11-14
- Initial commit

test.js
/*!
 * minibase-tests <https://github.com/node-minibase/minibase-tests>
 *
 * Copyright (c) Charlike Mike Reagent <@tunnckoCore> (http://www.tunnckocore.tk)
 * Released under the MIT license.
 */

/* jshint asi:true */

'use strict'

var test = require('mukla')
var suite = require('./index')
var Base = require('base')
var Assemble = require('assemble-core')
var Templates = require('templates')
var MiniBase = require('minibase').MiniBase

var version = process.version.match(/^.(\d+)\.(\d+)/)
var major = version[1]
var minor = version[2]

function MyApp (options) {
  if (!(this instanceof MyApp)) {
    return new MyApp(options)
  }
  this.options = { silent: true }
  MiniBase.call(this, options)
}
MiniBase.extend(MyApp)

function BaseApp () {
  if (!(this instanceof BaseApp)) {
    return new BaseApp()
  }
  Base.call(this)
}
Base.extend(BaseApp)

test('should expose a function', function (done) {
  test.strictEqual(typeof suite, 'function')
  done()
})

test('should suite(123) reject with TypeError if App not a function', function (done) {
  suite(123).catch(function (err) {
    test.strictEqual(err.name, 'TypeError')
    test.strictEqual(err.message, 'minibase-tests: expect `App` to be constructor')
    done()
  })
})

test('should given App and MiniBase itself pass all tests', function (done) {
  suite(MyApp).runTests().then(function (res) {
    test.strictEqual(res.length, 0)
    return suite(MiniBase).runTests().then(function (tests) {
      var len = tests.length
      test.strictEqual(len, 0)
      done()
    }, done)
  }, done)
})

test('should `base` and Base-based app pass most of these tests', function (done) {
  var opts = { isBase: true }
  suite(Base, opts).runTests()
    .then(function (res) {
      test.strictEqual(res.length, 0)
      return suite(BaseApp, opts).runTests()
    }, done)
    .then(function (passed) {
      test.strictEqual(passed.length, 0)
      done()
    }, done)
})

test('should Templates and AssembleCore apps pass all tests', function (done) {
  if (major > 0 || minor > 11) {
    suite(Templates, { isBase: true }).runTests().then(function (result) {
      var length = result.length
      test.strictEqual(length === 0, true)
      return suite(Assemble, { isBase: true }).runTests()
    }, done)
    .then(function (pass) {
      test.strictEqual(pass.length, 0)
      done()
    }, done)
    return
  }
  done()
})


runner.js
'use strict'

var each = require('each-promise')
var Bluebird = require('bluebird')

function Runner () {
  this._tests = []
}

Runner.prototype.addTest = function addTest (title, fn) {
  function wrapper () {
    var promise = new Bluebird(function (resolve) {
      resolve(fn())
    })
    return promise
  }
  wrapper.title = title
  this._tests.push(wrapper)
}

Runner.prototype.runTests = function runTests (show) {
  var titles = []
  return each.serial(this._tests, {
    settle: true,
    flat: false,
    Promise: Bluebird,
    beforeEach: function (item) {
      titles[item.index] = item.value.title
    },
    afterEach: function (item) {
      /* istanbul ignore next */
      if (!show) return

      /* istanbul ignore next */
      if (item.reason) {
        console.error('not ok', item.index + 1, titles[item.index])
        console.error(item.reason.stack)
        return
      }

      /* istanbul ignore next */
      console.log('ok', item.index + 1, titles[item.index])
    }
  }).then(function (res) {
    var result = res.map(function (item) {
      item.title = titles[item.index]
      /* istanbul ignore next */
      return item.reason ? item : false
    })
    result = result.filter(Boolean)
    result.tests = res.length
    return result
  })
}

module.exports = Runner


index.js
/*!
 * minibase-tests <https://github.com/node-minibase/minibase-tests>
 *
 * Copyright (c) Charlike Mike Reagent <@tunnckoCore> (http://www.tunnckocore.tk)
 * Released under the MIT license.
 */

'use strict'

var assert = require('assert')
var Bluebird = require('bluebird')
var assertKindof = require('assert-kindof')
var EventEmitter3 = require('eventemitter3')
var Runner = require('./runner')

/**
 * > Test any app based on [minibase][] and [base][], just
 * pass constructor as `App` argument. If it is `base` based
 * pass `opts.isBase: true` option. When run `.runTests` it
 * returns resolved Promise with array with length of 0 if all
 * tests are passed. If any of the tests fails that `result`
 * array will contain these tests - their title, index and the error.
 * Resolved array also has `.tests` property which is the count of
 * all tests, so easily you can do `res.tests - res.length` to find
 * how many tests are failed, and get them by outputing `res`.
 *
 * **Example**
 *
 * ```js
 * var suite = require('minibase-tests')
 * var Base = require('base')
 * var Assemble = require('assemble-core')
 * var Templates = require('templates')
 * var MiniBase = require('minibase').MiniBase
 *
 * suite(Base, { isBase: true })
 * .runTests().then(function (res) {
 *   // if `res` has bigger length
 *   // it will contain failed tests
 *   console.log(res.length) // => 0
 *   console.log(res.tests) // => 17
 * })
 * suite(Assemble, { isBase: true })
 * .runTests().then(function (res) {
 *   console.log(res.length) // => 0
 *   console.log(res.tests) // => 17
 * })
 * suite(Templates, { isBase: true })
 * .runTests().then(function (res) {
 *   console.log(res.length) // => 0
 *   console.log(res.tests) // => 17
 * })
 *
 * // MiniBase itself passes these tests too
 * suite(MiniBase).runTests().then(function (res) {
 *   console.log(res.length) // => 0
 *   console.log(res.tests) // => 17
 * })
 *
 * function MyApp () {
 *   MiniBase.call(this)
 * }
 * MiniBase.extend(MyApp)
 *
 * suite(MyApp).runTests().then(function (res) {
 *   console.log(res.length) // => 0
 *   console.log(res.tests) // => 17
 * })
 * ```
 *
 * @name   suite
 * @param  {Function} `App` app constructor, if not a function returns rejected promise
 * @param  {Object} `opts` optional object, pass `isBase: true` for [base][] apps
 * @return {Promise} promise if `App` not a function or instance of [Runner](./runner.js), so call `.runTests()`
 * @api public
 */

module.exports = function minibaseTests (App, opts) {
  if (typeof App !== 'function') {
    var err = new TypeError('minibase-tests: expect `App` to be constructor')
    return Bluebird.reject(err)
  }
  opts = opts && typeof opts === 'object' ? opts : {}
  var app = App()
  var runner = new Runner()

  runner.addTest('should return an instance of MiniBase with or without `new` keyword', function () {
    if (opts.isBase) {
      return Bluebird.resolve()
    }
    assert.strictEqual(typeof (new App()), 'object')
    assert.strictEqual(typeof (App()), 'object')
    return Bluebird.resolve()
  })

  runner.addTest('should be an instance of event emitter', function () {
    if (opts.isBase) {
      return Bluebird.resolve()
    }
    assert.strictEqual(App() instanceof EventEmitter3, true, 'should be instanceof EventEmitter3')
    assert.strictEqual((new App()) instanceof EventEmitter3, true, 'should be instanceof EventEmitter3')
    return Bluebird.resolve()
  })

  runner.addTest('should expose `.delegate`, `.define` and `.extend` static methods', function () {
    assertKindof.function(App.extend)
    if (opts.isBase) {
      return Bluebird.resolve()
    }
    assertKindof.function(App.delegate)
    assertKindof.function(App.define)
    return Bluebird.resolve()
  })

  runner.addTest('should expose `.define`, `.use` and `.delegate` prototype methods', function () {
    assert.ok(assertKindof.is.function(app.use))
    assert.ok(assertKindof.is.function(app.define))

    if (opts.isBase) {
      return Bluebird.resolve()
    }
    assert.ok(assertKindof.is.function(app.delegate))
    return Bluebird.resolve()
  })

  runner.addTest('should extend the given Ctor with static methods', function () {
    function Ctor () {
      App.call(this)
    }
    App.extend(Ctor)
    assertKindof.function(Ctor.extend)
    if (!opts.isBase) {
      assertKindof.function(Ctor.define)
      assertKindof.function(Ctor.delegate)
    }

    function foo () {}
    Ctor.extend(foo)
    assertKindof.function(foo.extend)

    if (!opts.isBase) {
      assert.strictEqual(typeof foo.delegate, 'function')
      assert.strictEqual(typeof foo.define, 'function')
    }

    assertKindof.undefined(foo())
    assertKindof.object(new Ctor())
    return Bluebird.resolve()
  })

  runner.addTest('should extend the prototype of given Ctor', function () {
    function Foo () {
      App.call(this)
    }
    App.extend(Foo)

    var foo = new Foo()
    assertKindof.function(foo.use)
    assertKindof.function(foo.define)
    if (!opts.isBase) assertKindof.function(foo.delegate)
    return Bluebird.resolve()
  })

  runner.addTest('should expose `options` instance properties', function () {
    assertKindof.object(app.options)
    if (!opts.isBase) assertKindof.number(app._anonymousPluginsCount)
    return Bluebird.resolve()
  })

  runner.addTest('should expose event emitter prototype methods', function () {
    var some = new App({
      foo: 'qqq'
    })
    assertKindof.function(some.on)
    assertKindof.object(some)
    assertKindof.object(some.options)
    assertKindof.function(some.emit)
    assertKindof.function(some.once)
    return Bluebird.resolve()
  })

  runner.addTest('should call the function passed to `.use` method', function () {
    return new Bluebird(function (resolve, reject) {
      var base = App()
      base.once('error', reject)
      base.use(function () {
        resolve()
      })
    })
  })

  runner.addTest('should `.use` has instance context', function () {
    return new Bluebird(function (resolve, reject) {
      var base = new App()
      base.foo = 123
      base.once('error', reject)
      base.use(function () {
        assert.strictEqual(this instanceof App, true)
        if (!opts.isBase) assert.strictEqual(this instanceof EventEmitter3, true)
        assert.strictEqual(this.foo, 123)
        resolve()
      })
    })
  })

  runner.addTest('should `.use` be passed with app instance context', function () {
    return new Bluebird(function (resolve) {
      app.bar = 'foo'
      app.use(function (self) {
        assert.strictEqual(self.bar, 'foo')
        assert.strictEqual(self instanceof App, true)
        if (!opts.isBase) assert.strictEqual(self instanceof EventEmitter3, true)
        resolve()
      })
    })
  })

  runner.addTest('should `.define` a key-value pair on the instance', function () {
    var app = new App({ aaa: 'bbb' })

    app.define('foo', 'bar')
    assert.strictEqual(app.foo, 'bar')
    return Bluebird.resolve()
  })

  runner.addTest('should `.define` an own property', function () {
    app.define('bar', 123)
    assert.strictEqual(app.hasOwnProperty('bar'), true)
    return Bluebird.resolve()
  })

  runner.addTest('should `.define` a non-emumerable property', function () {
    app.define('qux', 'bar')
    assert.strictEqual(Object.keys(app).indexOf('qux'), -1)
    return Bluebird.resolve()
  })

  runner.addTest('should define multiple properties with `.delegate`', function () {
    if (opts.isBase) {
      return Bluebird.resolve()
    }
    app.delegate({
      foo: 'xxx',
      baz: 'zzz'
    })

    assert.strictEqual(app.hasOwnProperty('foo'), true)
    assert.strictEqual(app.hasOwnProperty('baz'), true)
    assert.strictEqual(app.foo, 'xxx')
    assert.strictEqual(app.baz, 'zzz')
    return Bluebird.resolve()
  })

  runner.addTest('should emit `error` event when error in plugin', function () {
    if (opts.isBase) {
      return Bluebird.resolve()
    }
    var app = new App({ silent: true })

    return new Bluebird(function (resolve) {
      app.on('error', function (err) {
        assertKindof.error(err)
        assertKindof.string(err.name)
        assertKindof.string(err.message)
        assert.strictEqual(err.message, 'plugin err')
        assert.strictEqual(err instanceof Error, true)
        resolve()
      })
      app.use(function (app) {
        throw new Error('plugin err')
      })
    })
  })

  runner.addTest('should default MiniBase plugins passed to `.use` be synchronous', function () {
    return new Bluebird(function (resolve, reject) {
      app.once('error', reject)
      app.use(function (self, cb) {
        if (!opts.isBase) assert.strictEqual(arguments.length, 1)
        if (!opts.isBase) assert.strictEqual(assertKindof.is.function(cb), false)
        assertKindof.object(self)
        resolve()
      })
    })
  })

  return runner
}
